[FLINK-25899] Add connected components example

tillrohrmann opened the pull request:
This PR adds an example for calculating the connected components using Stateful Functions.

This PR needs to be ported to `dev` and `release-3.2`.

igalshilman reviewed (approved):
Thanks @tillrohrmann! this is a very interesting example, and really shows the versatility of StateFun! 

I've left few comments for your consideration, mainly it is about the use of Java serialisation, and `Type`s in general.

tillrohrmann commented on java/connected-components/src/main/java/org/apache/flink/statefun/playground/java/connectedcomponents/types/JavaType.java:
```diff
@@ -0,0 +1,70 @@
+    private static final class JavaSerializer<T> implements TypeSerializer<T> {
+        @Override
+        public Slice serialize(T value) {
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
Sounds good. Let's not give a wrong example. I'll remove the Java serialization.

tillrohrmann commented on java/connected-components/src/main/protobuf/messages.proto:
```diff
@@ -0,0 +1,30 @@
+option java_package = "org.apache.flink.statefun.playground.java.connectedcomponents.types.generated";
+option java_multiple_files = true;
+
+message VertexComponentChange {
```
I think this makes a lot of sense. I will remove the protobuf dependency.

tillrohrmann commented on java/connected-components/src/main/java/org/apache/flink/statefun/playground/java/connectedcomponents/types/Types.java:
```diff
@@ -0,0 +1,38 @@
+   */
+  public static final Type<Vertex> VERTEX_INIT_TYPE =
+      SimpleType.simpleImmutableTypeFrom(
+          TypeName.typeNameOf(TYPES_NAMESPACE, Vertex.class.getName()),
```
True. Yes, will use a string constant.

tillrohrmann commented:
Thanks for the review @igalshilman. I've addressed your comments and pushed an update. Merging this PR now into `release-3.1`, `release-3.2` and `dev`.
